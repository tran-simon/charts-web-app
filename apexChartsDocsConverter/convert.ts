export type DocValue = {
  name: string,
  type: string,
  description: string,
  options?: string[]
  path: string[]
  _isDocValue: true
}

type DocSection = {
  name: string,
  children: DocOption[],
  path: string[]
  _isDocSection: true
}

type DocOption = DocValue | DocSection

export const isDocValue = (option: DocValue | DocSection): option is DocValue=>{
  return typeof option === 'object' && '_isDocValue' in option;
}

export const isDocSection = (option: DocValue | DocSection): option is DocSection=>{
  return typeof option === 'object' && '_isDocSection' in option;
}

const mapOptions = (divs: any[], path: string[]):DocOption[]=>{
  return divs.map((entry) => {
    const children: any[] = entry?.children || []

    const h3 = children.find((v) => v.tagName === 'h3');
    const [name, type] = (h3?.children[0]?.content as string)?.split(': ') || []

    const p = children.find((v) => v.tagName === 'p');
    const description = p?.children?.find((v: any) => v.type === 'text')?.content?.trim();

    const ul = children.find((v) => v.tagName === 'ul')

    const options = ul?.children?.map((v: any) => {
      return v?.children[0]?.content
    })

    const divs = children?.filter((v) => v.tagName === 'div') || [];
    const newPath = [...path, name]
    if (divs.length) {
      return {
        name,
        children: mapOptions(divs, newPath),
        _isDocSection: true,
        path: newPath
      }
    }

    return {
      name,
      type: type?.toLowerCase(),
      description,
      options,
      _isDocValue: true,
      path: newPath
    };
  });
}

export default (jsonData: any, name: string) => {
  const list: any[] = jsonData[0]?.children;

  const mappedOptions = mapOptions(list.filter((o) => o.tagName === 'div'), [name])

  return mappedOptions.reduce((acc, curr)=>{
    acc[curr.name] = curr
    return acc
  }, {} as any)
}
