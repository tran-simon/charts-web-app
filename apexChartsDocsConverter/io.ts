import { promises as fs } from 'fs';
// @ts-ignore
import { parse } from 'himalaya';
import { JSDOM } from 'jsdom';
import { isDocSection, isDocValue } from './convert';
import prettier from 'prettier'

export const generateCode = async (data: string, name: string) => {
  const res = `
import { ApexOptions } from 'apexcharts';
import { Options } from '../optionModel';

const ${name}: Options<ApexOptions['${name}']> = ${data};

export default ${name}; 
`;

  const options = (await prettier.resolveConfig('../.prettierrc.json')) || {
    'semi': true,
    'trailingComma': 'all',
    'singleQuote': true,
    'tabWidth': 2,
  };
  return prettier.format(res, {
    parser: 'babel',
    ...options
  })
};

export const stringify = (data: object): string => {
  const replacer = (key: string, v: any) => {
    if (isDocValue(v)) {
      if (v.options) {

        const selectOptions = v.options
          .map((option) =>
            `${option}: { value: '${option}', labelId: ${[...v.path, 'option', option].join('.')} }`)

        return `new SelectOptionField({ ${selectOptions} })`;
      }

      if (v.type === 'string') {
        return `new TextOptionField()`;
      }
      if (v.type === 'number') {
        return `new NumberOptionField()`;
      }
      if (v.type === 'boolean') {
        return `new BoolOptionField()`;
      }
      if (v.type === 'string | number' || v.type === 'number | string') {
        // TODO(https://github.com/tran-simon/pretty-charts/issues/5)
        return `new TextOptionField()`;
      }
      if (v.type === 'color'){
        // TODO(https://github.com/tran-simon/pretty-charts/issues/6)
        return `new TextOptionField()`;
      }
    } else if (isDocSection(v)) {
      return  v.children.reduce((acc, curr) => {
        acc[curr.name] = curr
        return acc;
      }, {} as any);
    }
    return v;
  };

  return JSON.stringify(data, replacer, 2).replaceAll('"', '');
};

/**
 * HTML file to js object
 * @param html document data
 */
export const parseHtml = (html: Document) => {
  const optionsDetails = html.getElementsByClassName('options-details').item(0);

  if (!optionsDetails) {
    throw 'Error while retrieving HTML element: element with class `options-details` not found.';
  }

  return parse(optionsDetails.innerHTML);
};
