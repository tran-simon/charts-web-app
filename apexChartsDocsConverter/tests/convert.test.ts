import * as assert from 'assert';
import fs from 'fs';
import convert from '../convert';
import { parseHtml } from '../io';
import { JSDOM } from 'jsdom';

describe("convert", ()=>{
  let jsonData = {}

  beforeAll(async () => {
    const document = (await JSDOM.fromFile('tests/title.html')).window.document;
    jsonData = parseHtml(document);
  })

  it('can convert title.html', ()=>{
    expect(convert(jsonData, 'title')).toMatchSnapshot()
  })
})
