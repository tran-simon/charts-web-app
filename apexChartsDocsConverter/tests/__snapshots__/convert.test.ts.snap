// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert can convert title.html 1`] = `
Object {
  "align": Object {
    "_isDocValue": true,
    "description": "Alignment of title relative to chart area.",
    "name": "align",
    "options": Array [
      "left",
      "center",
      "right",
    ],
    "path": Array [
      "title",
      "align",
    ],
    "type": "string",
  },
  "floating": Object {
    "_isDocValue": true,
    "description": "The floating option will take out the title text from the chart area and make it float on top of the chart.",
    "name": "floating",
    "options": undefined,
    "path": Array [
      "title",
      "floating",
    ],
    "type": "boolean",
  },
  "margin": Object {
    "_isDocValue": true,
    "description": "Vertical spacing around the title text",
    "name": "margin",
    "options": undefined,
    "path": Array [
      "title",
      "margin",
    ],
    "type": "number",
  },
  "offsetX": Object {
    "_isDocValue": true,
    "description": "Sets the left offset for title text",
    "name": "offsetX",
    "options": undefined,
    "path": Array [
      "title",
      "offsetX",
    ],
    "type": "number",
  },
  "offsetY": Object {
    "_isDocValue": true,
    "description": "Sets the top offset for title text",
    "name": "offsetY",
    "options": undefined,
    "path": Array [
      "title",
      "offsetY",
    ],
    "type": "number",
  },
  "style": Object {
    "_isDocSection": true,
    "children": Array [
      Object {
        "_isDocValue": true,
        "description": "Font Size of the title text",
        "name": "fontSize",
        "options": undefined,
        "path": Array [
          "title",
          "style",
          "fontSize",
        ],
        "type": "string",
      },
      Object {
        "_isDocValue": true,
        "description": "Font Weight of the title text",
        "name": "fontWeight",
        "options": undefined,
        "path": Array [
          "title",
          "style",
          "fontWeight",
        ],
        "type": "string | number",
      },
      Object {
        "_isDocValue": true,
        "description": "Font Family of the title text",
        "name": "fontFamily",
        "options": undefined,
        "path": Array [
          "title",
          "style",
          "fontFamily",
        ],
        "type": "string",
      },
      Object {
        "_isDocValue": true,
        "description": "Fore color of the title text",
        "name": "color",
        "options": undefined,
        "path": Array [
          "title",
          "style",
          "color",
        ],
        "type": "color",
      },
    ],
    "name": "style",
    "path": Array [
      "title",
      "style",
    ],
  },
  "text": Object {
    "_isDocValue": true,
    "description": "Text to display as a title of chart",
    "name": "text",
    "options": undefined,
    "path": Array [
      "title",
      "text",
    ],
    "type": "string",
  },
}
`;
